src/Disks/DiskCache.cpp

```diff
@@ -0,0 +1,1121 @@
+                part.type = DiskCachePolicy::CachePart::CachePartType::ABSENT_NO_CACHE;
+            }
+
+            if (part.type == DiskCachePolicy::CachePart::CachePartType::EMPTY)
+            {
+                from = std::make_unique<ReadBufferFromEmptyFile>();
+                return;
+            }
+
+            if (part.type == DiskCachePolicy::CachePart::CachePartType::CACHED)
+            {
+                String cache_path = cache_base_path + "_" + std::to_string(part.offset);
+                LOG_TRACE(log, "ensureFrom try to read from cache: {}", cache_path);
+                cache_policy->read(cache_key, part.offset, part.size);
+                try
+                {
+                    size_t required_size = part.size;
+                    if (part.offset + part.size > current_offset + max_size)
+                        required_size = current_offset + max_size - part.offset;
+                    auto from_seekable = cache_disk->readFile(cache_path, ReadSettings(), required_size);
+                    if (current_offset > part.offset)
+                        from_seekable->seek(current_offset - part.offset, SEEK_SET);
+                    from = std::move(from_seekable);
+                    current_offset = part.offset + part.size;
+                }
+                catch (const Exception & e)
+                {
+                    if (e.code() == ErrorCodes::FILE_DOESNT_EXIST)
+                    {
+                        LOG_WARNING(log, "Cached file {} not found, try to download", cache_path);
+                        cache_policy->remove(cache_key, part.offset, part.size);
+                        --retries;
+                        continue; // retry
+                    }
+                    else
+                    {
+                        throw;
+                    }
+                }
+            }
+
+            if (part.type == DiskCachePolicy::CachePart::CachePartType::ABSENT)
+            {
+                LOG_TRACE(log, "ensureFrom download key {} from s3 {}+{}", cache_key, part.offset, part.size);
+                auto to_remove = cache_policy->reserve(part.size);
+                for (auto & entry_to_remove : to_remove)
+                {
+                    String remove_cache_path = DiskCache::getCacheBasePath(entry_to_remove.first) + "_" + std::to_string(entry_to_remove.second);
+                    LOG_TRACE(log, "Remove cached file {}", remove_cache_path);
+                    cache_disk->removeFileIfExists(remove_cache_path);
+                }
+
+                ProfileEvents::increment(ProfileEvents::DiskCacheRequestsOut, 1);
+                auto s3from = downloader->get(part.offset, part.size);
+                if (s3from.stream)
+                {
+                    String cache_path = cache_base_path + "_" + std::to_string(part.offset);
+                    from = std::unique_ptr<ReadBuffer>{ new CacheableReadBufferDecorator(s3from.stream,
```

Comment:
So am I correct now buffer chain is like this:

case `ABSENT`:
`(Asycnhronous)ReadIndirectBufferFromRemoteFS -> ReadBufferFromRemoteFSGather -> ReadBufferFromS3 -> CacheableMultipartReadBufferDecorator -> CacheableReadBufferDecorator -> ReadBufferFromIStream`

case `ABSENT_NO_CACHE`:
`(Asycnhronous)ReadIndirectBufferFromRemoteFS -> ReadBufferFromRemoteFSGather -> ReadBufferFromS3 -> CacheableMultipartReadBufferDecorator -> ReadBufferFromIStream`

case `CACHED`:
`(Asycnhronous)ReadIndirectBufferFromRemoteFS -> ReadBufferFromRemoteFSGather -> ReadBufferFromS3 -> CacheableMultipartReadBufferDecorator -> cache_disk->readFile`

May be it is better in these chains to move `ReadBufferFromS3` to the end? So it can be more easily (and  intuitively) applied to disks `hdfs`, `web` and soon `Azure blob storage`? Now this `DiskCache.cpp` file looks too much s3 specific.